compiler/rustc_hir_typeck/src/expr.rs

```diff
@@ -1330,7 +1330,11 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     ) -> Ty<'tcx> {
         let rcvr_t = self.check_expr(rcvr);
         // no need to check for bot/err -- callee does that
-        let rcvr_t = self.structurally_resolve_type(rcvr.span, rcvr_t);
+        let rcvr_t = if let ExprKind::Index(_, index, _) = rcvr.kind {
+            self.structurally_resolve_type(index.span, rcvr_t)
```

Comment:
This is not correct in case the ambiguity is not due to the index expression:

```rust
fn main() {
    let x = vec![];
    x[0usize].method();
}
```

```
error[E0282]: type annotations needed for `Vec<_>`
 --> /home/michael/test.rs:2:9
  |
2 |     let x = vec![];
  |         ^
3 |     x[0usize].method();
  |       ------ type must be known at this point
```

See above how after this PR, we now underline the `0usize` even though it is not ambiguous.

Again, I really don't think we should make this change in such a special case. 

Systematically determining *why* the method receiver is ambiguous due to inference is difficult, and I'd rather not patch a single case since it doesn't seem to me more special than any other example.

Reply:
How about the new change?

Reply:
This is not correct for cases where the key's ambiguity doesn't affect the value's ambiguity. Consider `HashMap<i32, _>`.